api/lib/opentelemetry/trace/span.rb

```diff
@@ -81,25 +81,37 @@ def add_event(name, attrs = nil)
         self
       end
 
-      # Adds a link to another Span from this Span. Linked Span can be from the
-      # same or different trace. See Links description.
+      # Adds a link to another Span from this Span. The linked Span can be from
+      # the same or different trace. See {Link} for a description.
       #
-      # @param [SpanContext, Link] span_context_or_link SpanContext of the Span
-      #   to link to or a Link which allows allows instrumentation to supply a
-      #   lazily calculated Link.
+      # @param [SpanContext] span_context SpanContext of the Span to link with
+      #   Span
       # @param [Hash<String, Object>] attrs Map of attributes associated with
       #   this link. Attributes are key:value pairs where key is a string and
       #   value is one of string, boolean or numeric type.
       #
       # @return [self] returns itself
-      def add_link(span_context_or_link, attrs = nil)
-        raise ArgumentError if span_context_or_link.nil?
-        raise ArgumentError unless span_context_or_link.instance_of?(SpanContext) || attrs.nil? || attrs.empty?
+      def add_link(span_context, attrs = nil)
+        raise ArgumentError if span_context.nil?
+        raise ArgumentError unless span_context.instance_of?(SpanContext) || attrs.nil? || attrs.empty?
         raise ArgumentError unless valid_attributes?(attrs)
 
         self
       end
 
+      # Adds a link to another Span from this Span. The linked Span can be from
+      # the same or different trace. See {Link} for a description.
+      #
+      # @param [Link] link A link to another span whose attributes are lazily
+      #   accessed
+      #
+      # @return [self] returns itself
+      def add_lazy_link(link)
+        raise ArgumentError if link.nil? || !link.is_a?(Link)
+
+        self
+      end
```

Comment:
The spec is pretty clear that we should have an API method that takes a `span_context` and attributes, and a method that takes a `Link` object.

> - An API to record a single Link where the Link properties are passed as arguments. This MAY be called AddLink.
> - An API to record a single lazily initialized Link. This can be implemented by providing a Link interface or a concrete Link definition and a LinkFormatter. If the language supports overloads then this MAY be called AddLink otherwise AddLazyLink MAY be consider.

I went ahead and combined `Span#add_link` and `Span#add_lazy_link`. The user can pass in a `span_context` with `attributes`, or an `attribute_formatter`, or just a link instance.